www/src/components/repository/RepositoryEdit.tsx

```diff
@@ -0,0 +1,677 @@
+
+import {
+  REPOSITORY_QUERY,
+  TAGS_SEARCH_QUERY,
+  UPDATE_REPOSITORY_MUTATION,
+} from './queries'
+
+export const categories = [
+  'DEVOPS',
+  'DATABASE',
+  'MESSAGING',
+  'SECURITY',
+  'DATA',
+  'PRODUCTIVITY',
+  'NETWORK',
+  'STORAGE',
+]
+
+const StyledTextInput = styled(TextInput)`
+  border: 1px solid #454954;
+  border-radius: 3px;
+  &:focus {
+    border-color: #5c77ff;
+  }
+`
+
+function useUpdateState<T extends { [key: string]: unknown }>(initialState: T) {
+  const [state, setState] = useState({ ...initialState })
+
+  const update = useCallback(
+    (update: Partial<T>) => {
+      setState({ ...state, ...update })
+    },
+    [state]
+  )
+  const reset = useCallback(() => {
+    setState({ ...initialState })
+  }, [initialState])
+
+  const hasUpdates = useMemo(() => {
+    for (const [prop, value] of Object.entries(state)) {
+      const curVal = !value && typeof value === 'string' ? null : value
+      const initialVal =
+        !initialState[prop] && typeof value === 'string'
+          ? null
+          : initialState[prop]
+      if (!isEqual(curVal, initialVal)) {
+        return true
+      }
+    }
+
+    return false
+  }, [initialState, state])
+
+  return {
+    state: { ...state },
+    hasUpdates,
+    update,
+    reset,
+    initialState: { ...initialState },
+  }
+}
+
+function RepoIcon({
+  src = null,
+  onClick,
+}: {
+  src: string | null;
+  onClick: EventHandler<any>;
+}) {
+  if (!src) {
+    return (
+      <Flex
+        width={96}
+        height={96}
+        backgroundColor="fill-two"
+        _hover={{
+          backgroundColor: 'fill-two-hover',
+        }}
+        border="1px solid border-fill-two"
+        borderRadius="medium"
+        align="center"
+        justify="center"
+        cursor="pointer"
+        onClick={onClick}
+      >
+        <P
+          title2
+          color="text-light"
+        >
+          +
+        </P>
+      </Flex>
+    )
+  }
+
+  return (
+    <Img
+      src={src}
+      alt="Icon"
+      width={64 + 32}
+      height={64 + 32}
+      objectFit="cover"
+      backgroundColor="fill-one"
+      cursor="pointer"
+      border="1px solid border"
+      onClick={onClick}
+    />
+  )
+}
+
+type FormState = {
+  name: string;
+  description: string;
+  category: string;
+  oauthUrl: string;
+  oauthMethod: string;
+  tags: { name: string }[];
+  private: boolean;
+};
+
+function RepositoryEdit() {
+  const {
+    id,
+    name,
+    description,
+    category,
+    editable,
+    oauthSettings,
+    tags,
+    icon,
+    private: privateRepo,
+  } = useContext(RepositoryContext) as any
+  const {
+    state: formState,
+    initialState: formInitialState,
+    hasUpdates: formStateHasUpdates,
+    update: updateFormState,
+    // reset: resetFormState,
+  } = useUpdateState<FormState>(
+    useMemo(
+      () => ({
+        name,
+        description,
+        category,
+        oauthUrl:
+          typeof oauthSettings?.uriFormat === 'string'
```

Comment:
couldn't this just be `oauthSettings?.uriFormat || ''`?  Or is the typeof clause doing some work to hint ts to it being a string

Reply:
We can also use `${oauthSettings?.uriFormat || ''}` to make sure that it is a string.

Reply:
To make sure that it is a string we can use:

```
`${oauthSettings?.uriFormat || ''}`
```